Source/WTF/wtf/text/WYHasher.h

```diff
@@ -0,0 +1,264 @@
+    }
+
+    ALWAYS_INLINE static constexpr uint64_t wyrot(uint64_t x)
+    {
+        return (x >> 32) | (x << 32);
+    }
+
+    ALWAYS_INLINE static constexpr void wymum(uint64_t* A, uint64_t* B)
+    {
+#if defined(__SIZEOF_INT128__)
+        __uint128_t r = *A;
+        r *= *B;
+        *A = static_cast<uint64_t>(r);
+        *B = static_cast<uint64_t>(r >> 64);
+#else
+        uint64_t ha = *A >> 32;
+        uint64_t hb = *B >> 32;
+        uint64_t la = static_cast<uint32_t>(*A);
+        uint64_t lb = static_cast<uint32_t>(*B);
+        uint64_t hi;
+        uint64_t lo;
+        uint64_t rh = ha * hb;
+        uint64_t rm0 = ha * lb;
+        uint64_t rm1 = hb * la;
+        uint64_t rl = la * lb;
+        uint64_t t = rl + (rm0 << 32);
+        uint64_t c = t < rl;
+
+        lo = t + (rm1 << 32);
+        c += lo < t;
+        hi = rh + (rm0 >> 32) + (rm1 >> 32) + c;
+        *A = lo;
+        *B = hi;
+#endif
+    }
+
+    ALWAYS_INLINE static constexpr uint64_t wymix(uint64_t A, uint64_t B)
+    {
+        wymum(&A, &B);
+        return A ^ B;
+    }
+
+#define enable_is_constant_evaluated 0
+
+    template<typename T>
+    struct Reader16Bit {
+        ALWAYS_INLINE static constexpr uint64_t wyr2(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
+            uint64_t firstByte = static_cast<uint8_t>(p[0]);
+            uint64_t secondByte = static_cast<uint8_t>(p[0] >> 8);
+            return (firstByte << 16) | (secondByte << 8) | secondByte;
+        }
+
+        ALWAYS_INLINE static constexpr uint64_t wyr4(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
+            if (std::is_constant_evaluated() || enable_is_constant_evaluated) {
+                return static_cast<uint32_t>(p[0])
+                    | (static_cast<uint32_t>(p[1]) << 16);
+            }
+            uint32_t v;
+            memcpy(&v, p, 4);
+            return v;
+        }
+
+        ALWAYS_INLINE static constexpr uint64_t wyr8(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
+            if (std::is_constant_evaluated() || enable_is_constant_evaluated) {
+                return static_cast<uint64_t>(p[0])
+                    | (static_cast<uint64_t>(p[1]) << 16)
+                    | (static_cast<uint64_t>(p[2]) << 32)
+                    | (static_cast<uint64_t>(p[3]) << 48);
+            }
+            uint64_t v;
+            memcpy(&v, p, 8);
+            return v;
+    }
+    };
+
+    template<typename T>
+    struct Reader8Bit {
+        ALWAYS_INLINE static constexpr uint64_t wyr2(const T* p)
+        {
+            ASSERT(sizeof(T) == 1);
+            return static_cast<uint64_t>(p[0]) << 16;
+        }
+
+        ALWAYS_INLINE static constexpr uint64_t wyr4(const T* p)
+        {
+            ASSERT(sizeof(T) == 1);
+            if (std::is_constant_evaluated() || enable_is_constant_evaluated)
+                return static_cast<uint32_t>(p[0]) | (static_cast<uint32_t>(p[1]) << 16);
```

Comment:
Let's wrap it with `static_cast<uint8_t>()` to avoid accidental sign extension.
```
return static_cast<uint32_t>(static_cast<uint8_t>(p[0])) | (static_cast<uint32_t>(static_cast<uint8_t>(p[1])) << 16);
```